File: common/canon.ts
export function canonicalize(value: unknown): string {
  const seen = new WeakSet<object>();
  const norm = (v: any): any => {
    if (v === null || typeof v !== "object") return v;
    if (seen.has(v)) throw new Error("Circular structure not allowed");
    seen.add(v);
    if (Array.isArray(v)) return v.map(norm);
    const keys = Object.keys(v).sort();
    const out: any = {};
    for (const k of keys) out[k] = norm(v[k]);
    return out;
  };
  return JSON.stringify(norm(value));
}

File: common/canon.py
import json
from typing import Any

def canonicalize(value: Any) -> str:
    def norm(v):
        if v is None or isinstance(v, (str, int, float, bool)):
            return v
        if isinstance(v, list):
            return [norm(x) for x in v]
        if isinstance(v, dict):
            return {k: norm(v[k]) for k in sorted(v.keys())}
        raise TypeError(f"Unsupported type: {type(v)}")

    return json.dumps(norm(value), separators=(",", ":"), ensure_ascii=False)
